processing/src/main/java/io/druid/query/topn/TopNLexicographicResultBuilder.java

```diff
@@ -122,9 +183,19 @@ public int compare(
               DimValHolder o2
           )
           {
-            return comparator.compare(o1.getDimName(), o2.getDimName());
+            return comparator.compare(o2.getDimName(), o1.getDimName());
           }
         }
-    ).maximumSize(threshold).create();
+    );
   }
+
+  private boolean shouldAdd(String dimName)
+  {
+    final boolean belowThreshold = pQueue.size() < this.threshold;
+    final boolean belowMax = belowThreshold
+                             || this.comparator.compare(pQueue.peek().getTopNMetricVal(), dimName) < 0;
+    // Only add if dimName is after previousStop
+    return belowMax && comparator.compare(dimName, previousStop) > 0;
```

Comment:
is it possible to optimize the case where previousStop is null and shortcut the comparison altogether? Seems expensive to call comparator.compare every time?